django/src/rdwatch/tasks/__init__.py

```diff
@@ -1,6 +1,275 @@
+
+
+def get_closest_capture(
+    bbox: tuple[float, float, float, float],
+    timestamp: datetime,
+    constellation: str,
+    worldView=False,
+):
+    timebuffer = timedelta(days=1)
+    captures = None
+    if worldView:
+        captures = get_worldview_captures(timestamp, bbox, timebuffer)
+    else:
+        captures = list(get_bands(constellation, timestamp, bbox, timebuffer))
+
+        # Filter bands by requested processing level and spectrum
+        tempCaptures = []
+        for band in captures:
+            if band.level.slug == '2A' and band.spectrum.slug == 'visual':
+                tempCaptures.append(band)
+        captures = tempCaptures
+    # logger.warning(f'Captures: {len(captures)}')
+    if not captures:
+        return None
+    closest_capture = min(captures, key=lambda band: abs(band.timestamp - timestamp))
+
+    return closest_capture
+
+
+def get_range_captures(
+    bbox: tuple[float, float, float, float],
+    timestamp: datetime,
+    constellation: str,
+    timebuffer: timedelta,
+    worldView=False,
+):
+    captures = None
+    if worldView:
+        captures = get_worldview_captures(timestamp, bbox, timebuffer)
+    else:
+        captures = list(get_bands(constellation, timestamp, bbox, timebuffer))
+
+        # Filter bands by requested processing level and spectrum
+        tempCaptures = []
+        for band in captures:
+            if band.level.slug == '2A' and band.spectrum.slug == 'visual':
+                tempCaptures.append(band)
+        captures = tempCaptures
+    # logger.warning(f'Captures: {len(captures)}')
+    if not captures:
+        return None
+
+    return captures
+
+
+def fetch_boundbox_image(
+    bbox: tuple[float, float, float, float],
+    timestamp: datetime,
+    constellation: str,
+    worldView=False,
+):
+    capture = get_closest_capture(bbox, timestamp, constellation, worldView)
+    # logger.warning(f'Closest Capture is: {capture}')
+    if capture is None:
+        return None
+    bytes = None
+    if worldView:
+        bytes = get_worldview_processed_visual_bbox(capture, bbox)
+    else:
+        bytes = get_raster_bbox(capture.uri, bbox)
+    return {
+        'bytes': bytes,
+        'cloudcover': capture.cloudcover,
+        'timestamp': capture.timestamp,
+    }
 
 
 @shared_task
-def test_task(input_value: int) -> None:
-    print(f'Testing an async task! {input_value}')
+def get_siteobservations_images(
+    site_eval_id: int, baseConstellation='WV', force=False
+) -> None:
+    site_observations = SiteObservation.objects.filter(siteeval=site_eval_id)
+    transformer = Transformer.from_crs('EPSG:3857', 'EPSG:4326')
+    found_timestamps = {}
+    min_time = datetime.max
+    max_time = datetime.min
+    max_bbox = [float('inf'), float('inf'), float('-inf'), float('-inf')]
+    matchConstellation = ''
+    baseSiteEval = None
+    # First we gather all images that match observations
+    for observation in site_observations:
```

Comment:
I think we should use `.iterator()` on this queryset 
https://docs.djangoproject.com/en/4.2/ref/models/querysets/#iterator
```python
    for observation in site_observations.iterator():
```

There's no max limit on the number of observations in a site, and I believe i've seen some with thousands; a site with thousands or tens of thousands of observations, each with its own complex geojson (multi)polygon, could blow up the memory on our worker.

Reply:
good call